### ISSUE TITLE
segment fault in JavascriptArray::EntryPushJavascriptArrayNoFastPath()

### ISSUE THREAD

**Issue by sunlili** [microsoft/ChakraCore#6510]
Hi, running following code in ch (1.11.22, debug), the program will be terminated with segment fault.
```
var arr = [1, 2, 3, 4, 5];  
Array.prototype.push.call(arr);
```
output:
```
FATAL ERROR: ch.exe failed due to exception code c0000005
```
crash reason:
`arr` is a `Js::JavascriptCopyOnAccessNativeIntArray` instance, the value of `head` field is `1`, which is the index to `CacheForCopyOnAccessArraySegments`. In `JavascriptArray::EntryPushJavascriptArrayNoFastPath()`, `JavascriptNativeIntArray::ValidateArray()` will be called, which will use `head` as a pointer. This is a type confusion bug  and trigger segment fault.

ISec Lab
2020.9.27

**Comment by ppenzin** [comment]
I can't reproduce this on either Windows or Linux.

**Comment by sunlili** [comment]
@ppenzin do you use 1.11.22 debug version to run this testcase? It can only reproduce on debug version. I have reconfirmed that this crash does exists. Or you may run it on master, perhaps this bug has been fixed.
![scr1 (2)](https://user-images.githubusercontent.com/10752186/98315085-50153780-2012-11eb-9298-e48e3dff5fe8.png)
![scr1 (1)](https://user-images.githubusercontent.com/10752186/98315088-51defb00-2012-11eb-8439-706523505713.png)

**Comment by ppenzin** [comment]
OK, I can confirm the issue exists in 1.11 and is fixed in master.

I've closed the bug, since nothing is going to 1.11 aside from security fixes.

**Comment by sunlili** [comment]
OK, thanks a lot :)
